src/components/state-lookup.ts

```diff
@@ -0,0 +1,188 @@
+     * queried at a later date. This component will perform network requests to
+     * fetch the current state for a given room ID. It relies on
+     * {@link StateLookup#onEvent} being called with later events in order to
+     * stay up-to-date. This should be connected to the <code>onEvent</code>
+     * handler on the {@link Bridge}.
+     * @constructor
+     * @param {Object} opts Options for this constructor
+     * @param {MatrixClient} opts.client Required. The client which will perform
+     * /state requests.
+     * @param {string[]} opts.eventTypes The state event types to track.
+     * @throws if there is no client.
+     */
+    constructor (opts: StateLookupOpts) {
+        if (!opts.client) {
+            throw new Error("client property must be supplied");
+        }
+
+        this.lookupQueue = new PQueue({
+            concurrency: opts.stateLookupConcurrency || DEFAULT_STATE_CONCURRENCY,
+        });
+
+        this._client = opts.client;
+        (opts.eventTypes || []).forEach((t) => {
+            this.eventTypes[t] = true;
+        });
+    }
+
+    /**
+     * Get a stored state event.
+     * @param {string} roomId
+     * @param {string} eventType
+     * @param {string=} stateKey If specified, this function will return either
+     * the event or null. If not specified, this function will always return an
+     * array of events, which may be empty.
+     * @return {?Object|Object[]}
+     */
+    public getState(roomId: string, eventType: string, stateKey?: string): unknown|unknown[] {
+        const r = this.dict[roomId];
+        if (!r) {
+            return stateKey ? null : [];
+        }
+        const es = r.events;
+        if (!es[eventType]) {
+            return stateKey ? null : [];
+        }
+        if (stateKey) {
+            return es[eventType][stateKey] || null;
+        }
+    
+        return Object.keys(es[eventType]).map(function(skey) {
+            return es[eventType][skey];
+        });
+    }
+
+    /**
+     * Track a given room. The client must have access to this room.
+     *
+     * This will perform a room state query initially. Subsequent calls will do
+     * nothing, as it will rely on events being pushed to it via {@link StateLookup#onEvent}.
+     *
+     * @param {string} roomId The room ID to start tracking. You can track multiple
+     * rooms by calling this function multiple times with different room IDs.
+     * @return {Promise} Resolves when the room is being tracked. Rejects if the room
+     * cannot be tracked.
+     */
+    public trackRoom(roomId: string) {
+        const r = this.dict[roomId] = this.dict[roomId] || {};
+        if (r.syncPromise) {
+            return r.syncPromise;
+        }
+        r.events = {};
+        r.syncPromise = (async () => {
+            while (true) {
+                try {
+                    const events: StateLookupEvent[] = await this.lookupQueue.add(
+                        () => this._client.roomState(roomId)
```

Comment:
A returning arrow function inside
a function that is awaited inside
a try block inside
a while(true) block inside
an async function that is directly called inside
a Bluebird.resolve().
What does that even give me?

The complexity:
![giphub](https://media0.giphy.com/media/tPKoWQJk3cEbC/giphy.gif)